### ISSUE TITLE
TTV.LOL leaking userID with IP

### ISSUE THREAD

**Issue by ghost** [TTV-LOL/extensions#8]
As posted on Reddit:

"Just a warning to those using TTV LOL, especially if you are a streamer. The extension currently leaks your Twitch user ID and personal IP to their Russian proxy. For obvious reasons this could be a problem. I have pasted an example below of the redirect URL used.

api.ttv.lol/playlist/streamer.m3u8%3Fallow_source%3Dtrue%26fast_bread%3Dtrue%26p%3A2255857%26play_session_id%3Ad453f480gg1r99h45a3095366ca5363a%26player_backend%3Dmediaplayer%26playlist_include_framerate%3Dtrue%26reassignments_supported%3Dtrue%26sig%3D86df4f12ab1d2ba8d99f82a81680be74e549c67a%26supported_codecs%3Dvp09%252Cavc1%26token%3D%257B%2522adblock%2522%253Afalse%252C%2522authorization%2522%253A%257B%2522forbidden%2522%253Afalse%252C%2522reason%2522%253A%2522%2522%257D%252C%2522blackout_enabled%2522%253Afalse%252C%2522channel%2522%253A%2522streamer%2522%252C%2522channel_id%2522%253A44338537%252C%2522chansub%2522%253A%257B%2522restricted_bitrates%2522%253A%255B%255D%252C%2522view_until%2522%253A1924905600%257D%252C%2522ci_gb%2522%253Afalse%252C%2522geoblock_reason%2522%253A%2522%2522%252C%2522device_id%2522%253A%2522Ob6oHtGYHjIKKJjjjRDwMe%2522%252C%2522expires%2522%253A1620467116%252C%2522extended_history_allowed%2522%253Afalse%252C%2522game%2522%253A%2522%2522%252C%2522hide_ads%2522%253Afalse%252C%2522https_required%2522%253Atrue%252C%2522mature%2522%253Afalse%252C%2522partner%2522%253Afalse%252C%2522platform%2522%253A%2522web%2522%252C%2522player_type%2522%253A%2522site%2522%252C%2522private%2522%253A%257B%2522allowed_to_view%2522%253Atrue%257D%252C%2522privileged%2522%253Afalse%252C%2522role%2522%253A%2522%2522%252C%2522server_ads%2522%253Atrue%252C%2522show_ads%2522%253Atrue%252C%2522subscriber%2522%253Afalse%252C%2522turbo%2522%253Afalse%252C%2522user_id%2522%253D7665776918%252C%2522user_ip%2522%253A%2522142.36.78.245%2522%252C%2522version%2522%253A2%257D%26cdm%3Dwv%26player_version%3D1.4.0

Hopefully the devs will remove this information from it's extension, as it shouldn't be needed. Not trying to scare people away as I'm sure it was just easier for them to implement it this way but still, it is a security risk.

This does not apply to Purple Ad-Block, which only sends the channel name."

Tested this myself and it does leak that information. Haven't tested purple ad block though.

If you could make it so that only the channel name or vod is is sent, that would fix it.

**Comment by captainswain** [comment]
Hi,

I just want to clear up any confusion on this topic. As per our privacy policy, we are not storing your twitch user ids or IP from the request. The endpoint you are hitting is also not some random proxy in Russia. Depending on where you are located you will be hitting a server in new york or Germany where your request is parsed and then proxied. Once the handler call is complete your request falls out of scope and in turn, any identifying information in memory is gone. Once again this data is not stored and is not sent as part of the proxied request. 

It was implemented this way for a couple of reasons:

1. easier to just pull all query parameters
2. We pass your non-identifying arguments as part of the proxied request ie: fast_bread, supported_codecs, etc.


I personally don't have time to refactor the server and plugin currently, but I can put that on our roadmap as a feature. I can see how seeing your User ID is scary, but rest assured we are not storing it or passing it along in our proxied request.

**Comment by ghost** [comment]
Is this the correct privacy policy? https://addons.mozilla.org/en-US/firefox/addon/ttv-lol/privacy/ , because it does state you do not collect any personal information. A user id linked to a Twitch account and dedicated IP, would essentially give you that streamers address. I think the privacy policy should state that user id's are collected, but not stored, like you said.

You have to understand, that although the extension code is opensource, your server code is not, so we only have your word to go by.

I understand re-factoring your server-side code would take time, but a much faster way would be to simply replace the user id from the redirect url in the background.js file of the extension with either 0's or null. This could be done with just one line of code and would stop the issue, while also not affecting your server-side code.

**Comment by captainswain** [comment]
Yes I'm referring to that privacy policy. That is correct we do not collect any personally identifying information, collecting such data would imply storing that information in a collection.

I personally don't agree with "_A user id linked to a Twitch account and dedicated IP, would essentially give you that streamers address."_. That seems a little outlandish and you are blowing the severity of this a bit out of proportion. An IP address is not anywhere close to a geographic location and by itself don't serve as grounds for user identification. On top of that you can not pull any information from the [Twitch API endpoints](https://dev.twitch.tv/docs/v5/reference/users) that anybody else on the platform couldn't. Realistically, if we were bad actors all we could do with your user id and IP is send denial of service attacks or attempt some sort of social engineering attack against your ISP.

Luckily this project is Opensource so, please if you have the time work on a pull request and I'll review the code. As I said there is no ill intent and at some point it will be fixed.  I'd love for your contribution so we can end this Witchhunt that has been created surrounding the project.

**Comment by ghost** [comment]
Collecting does not imply storing the information, there is a difference in law between the two words. For example, Facebook may collect information and pass it on to another entity, without storing it. I'm only trying to help. Some Twitch streamers have business internet with dedicated IP's that do in fact have a home location, so you can link the two and it has been done before.

Also your browser extension is opensource, but your server code is not. The extension redirects the URL only, As users we have no idea what is happening server side because it is not opensource, hence the importance on getting this fixed asap.

I don't think you are doing anything malicious, but it's better to be 100% sure by just simply changing the userID.

As I don't know the encoding of the redirectURL I can't give you the exact code, but something like this would change the userID and IP to 0's.

`var redirectURL = "https://api.ttv.lol/playlist/streamer.m3u8%3Fallow_source%3Dtrue%26fast_bread%3Dtrue%26p%3D5455440%26play_session_id%3Df0dcc5656787ecce3ed0ecba287aa354%26player_backend%3Dmediaplayer%26playlist_include_framerate%3Dtrue%26reassignments_supported%3Dtrue%26sig%3D4e385054545gffggffggf28b251f08e201c62236d926%26supported_codecs%3Dvp09%252Cavc1%26token%3D%257B%2522adblock%2522%253Afalse%252C%2522authorization%2522%253A%257B%2522forbidden%2522%253Afalse%252C%2522reason%2522%253A%2522%2522%257D%252C%2522blackout_enabled%2522%253Afalse%252C%2522channel%2522%253A%2522streamer%2522%252C%2522channel_id%2522%253A4856561737%252C%2522chansub%2522%253A%257B%2522restricted_bitrates%2522%253A%255B%255D%252C%2522view_until%2522%253A1924905600%257D%252C%2522ci_gb%2522%253Afalse%252C%2522geoblock_reason%2522%253A%2522%2522%252C%2522device_id%2522%253A%2522Objghgffghg55656hFDyYFcBECRDwMe%2522%252C%2522expires%2522%253A1621024296%252C%2522extended_history_allowed%2522%253Afalse%252C%2522game%2522%253A%2522%2522%252C%2522hide_ads%2522%253Afalse%252C%2522https_required%2522%253Atrue%252C%2522mature%2522%253Afalse%252C%2522partner%2522%253Afalse%252C%2522platform%2522%253A%2522web%2522%252C%2522player_type%2522%253A%2522site%2522%252C%2522private%2522%253A%257B%2522allowed_to_view%2522%253Atrue%257D%252C%2522privileged%2522%253Afalse%252C%2522role%2522%253A%2522%2522%252C%2522server_ads%2522%253Atrue%252C%2522show_ads%2522%253Atrue%252C%2522subscriber%2522%253Afalse%252C%2522turbo%2522%253Afalse%252C%2522user_id%2522%253A12345678%252C%2522user_ip%2522%253A%2522123.456.789.000%2522%252C%2522version%2522%253A2%257D%26cdm%3Dwv%26player_version%3D1.4.0";

var searchTerm = 'user_id';
var indexOfFirst = redirectURL.indexOf(searchTerm);
var newRedirectURL = redirectURL.substring(0,indexOfFirst) + "user_id%2522%253A00000000%252C%2522user_ip%2522%253A%2522000.000.000.000%2522%252C%2522version%2522%253A2%257D%26cdm%3Dwv%26player_version%3D1.4.0";`

But I'm sure there is easier ways with RegEx.

**Comment by liamengland1** [comment]
Then the signature doesn't match. So that doesn't work.

**Comment by ghost** [comment]
The signature of the redirect URL does not matter. It is only the query params that the dev wanted, such as fast_bread, supported_codecs, etc. Their server would then make a new GQL request, giving a new token/sig for the final proxy server. You can see this by how the play_session_id is different in the final m3u8 response compared to the redirect url.

Ideally, the redirect url would be something like: https://api.ttv.lol/playlist/streamer&fastbread=true&support_codecs=vp09,avc  similar to how purple ad block has just the /channelname only, however the dev explained they did not have time to refactor the server code. So the faster alternative is to refactor the extension in this way, to fix the issue while not affecting their server code.

**Comment by liamengland1** [comment]
ok, I see. thanks for explaining.

**Comment by ghost** [comment]
@captainswain Could you confirm if my personal IP is sent to the final proxy? After looking at the m3u8 response that the extension gives, it contains my personal IP but also says USER-COUNTRY="RU" (I'm not in Russia). As earier you said "Once again this data is not stored and is not sent as part of the proxied request." Honestely just interested if you do actually make a new GQL request, or if it is the whole usher URL that you proxy, in which case refactoring the extension would be useless. Thanks

**Comment by ChoosenEye** [comment]
They are replacing the IP from their Server with your IP currently thats why its in the m3u8, in my Case i just replaced my Server IPs in the m3u8 with 127.0.0.1 which would be the easier and imo better way.

**Comment by ghost** [comment]
Ah got ya, ok that makes sense to prevent the IP leak to Twitch. Was worried, that it was the whole usher request being proxied, in which case they would never be able to fix the userID issue.

**Comment by austinmccalley** [comment]
I just opened a pull request(#10) that removes the private information which people are worried about.

**Comment by pixeltris** [comment]
Here's a couple of functions (they both do the same thing) which might be a bit more versatile.

The first completely strips `token` / `sig` as you likely don't use these at all in your server code. If you do, then the second function might be more suitable which just strips `user_id` from `token`.

```javascript
// Strips unused search params 'token' / 'sig'
function stripUnusedParams(str, params) {
    if (!params) {
        params = [ 'token', 'sig' ];
    }
    var tempUrl = new URL('https://localhost/' + str);
    for (var i = 0; i < params.length; i++) {
        tempUrl.searchParams.delete(params[i]);
    }
    return tempUrl.pathname.substring(1) + tempUrl.search;
}
```

```javascript
// Strips sensitive values in the 'token' param
function stripTokenInfo(str, tokenValues) {
    if (!tokenValues) {
        tokenValues = [ 'user_id' ];
    }
    var tempUrl = new URL('https://localhost/' + str);
    if (tempUrl.searchParams.has('token')) {
        var tokenInfo = JSON.parse(tempUrl.searchParams.get('token'));
        for (var i = 0; i < tokenValues.length; i++) {
            delete tokenInfo[tokenValues[i]];
        }
        tempUrl.searchParams.set('token', JSON.stringify(tokenInfo));
    }
    return tempUrl.pathname.substring(1) + tempUrl.search;
}
```

Example usage;

```javascript
redirectUrl: `https://api.ttv.lol/${playlistType}/${encodeURIComponent(stripUnusedParams(match[2]))}`,
```

Output:

```
https://api.ttv.lol/playlist/dashducks.m3u8%3Fallow_source%3Dtrue%26fast_bread%3Dtrue%26p%3DREDACTED%26play_session_id%3DREDACTED%26player_backend%3Dmediaplayer%26playlist_include_framerate%3Dtrue%26reassignments_supported%3Dtrue%26supported_codecs%3Dvp09%252Cavc1%26cdm%3Dwv%26player_version%3D1.4.0
```

**Comment by ghost** [comment]
Very nice solution here, hope it gets implemented asap :)

**Comment by saucettv** [comment]
@captainswain Do you have any spare time to implement this at the moment? Or could you give us a rough time frame? From personal experrience, once you have the Firefox extension approved, further changes are approved almost instantly. Chrome takes a few days though.

**Comment by ghost** [comment]
Ok, did the pull request for you, tested in Firefox and it's working. https://github.com/TTV-LOL/extensions/pull/11 , just need you to approve it and upload the new zips to Firefox and Chrome.

**Comment by ghost** [comment]
@captainswain I know it's only been 2 days since the pull request, but any chance you could give us a rough idea of when you would be able to complete it? From personal experience, Firefox updates extensions almost instantly after the first approval, Chrome still takes a few days depending on how much has changed. Also, I'm not sure if you are aware, but Firefox allow you to use chrome.webRequest.onBeforeRequest, etc. It makes it a alot easier, as you only need to keep one code base for both browsers.

**Comment by ghost** [comment]
So should I be concerned about this? Does this mean a malicious actor could obtain this information by attacking the proxies server? Because I was able to check this for myself and I did see my IP and a "user ID" field with random numbers.

**Comment by ghost** [comment]
Well it's been 11 days since we provided the fix and we have had no reply from the dev, I'm not sure why tbh. Although the dev seems genuine, they could technically link an IP to a username and worst case scenario is that user is using a business IP that links to a home address.

**Comment by ksthiele** [comment]
Who cares? IPs are public. What do you think they could do with it? Whenever you visit a website 20 services also get it.

**Comment by ghost** [comment]
The problem isn't the IP going to the external server. Like you said all IP's are sent to external servers. It's the fact that the IP and the Twitch userID/userName + IP is sent to a third party external server (When it isn't actually needed). This is personal information when linked like that. If you are a user and or streamer that happens to have a dedicated IP/business IP, you should care, especially if those IP's happen to be linked to your home address. I know plenty of streamers with a business IP linked to a home address. As the server code is not opensource, we are going by the word of the dev, that they are not collecting and correlating this information. This is a privacy issue that has already had the pull request sent to them that has the fix and can be updated within an hour. Yet there has been zero response from the dev on this issue for 21 days. As someone that remembers the swatting days and streamers being stalked/ DDoS'd, this is a fairly serious privacy issue.

**Comment by ghost** [comment]
> This is a privacy issue that has already had the pull request sent to them that has the fix and can be updated within an hour. Yet there has been zero response from the dev on this issue for 21 days. 

That is a little weird. Should I stop using it till he decides to update it? What difference would it make now if they already have that info right?

**Comment by ghost** [comment]
If you are a normal user with a normal IP, I wouldn't worry too much about it and like you said, if anything dodgy was happening, it's already too late. It's just a privacy issue that in my opinion should of been fixed a while ago now.

**Comment by captainswain** [comment]
I have merged the PR to fix this. It will be rolled out once I can test and submit the updated plugins to chrome & firefox.



> > This is a privacy issue that has already had the pull request sent to them that has the fix and can be updated within an hour. Yet there has been zero response from the dev on this issue for 21 days.
> 
> That is a little weird. Should I stop using it till he decides to update it? What difference would it make now if they already have that info right?

There was no real threat or disclosure of personal information. 




> The problem isn't the IP going to the external server. Like you said all IP's are sent to external servers. It's the fact that the IP and the Twitch userID/userName + IP is sent to a third party external server (When it isn't actually needed). This is personal information when linked like that. If you are a user and or streamer that happens to have a dedicated IP/business IP, you should care, especially if those IP's happen to be linked to your home address. I know plenty of streamers with a business IP linked to a home address. As the server code is not opensource, we are going by the word of the dev, that they are not collecting and correlating this information. This is a privacy issue that has already had the pull request sent to them that has the fix and can be updated within an hour. Yet there has been zero response from the dev on this issue for 21 days. As someone that remembers the swatting days and streamers being stalked/ DDoS'd, this is a fairly serious privacy issue.

I appreciate do appreciate @saucettv writing the PR, however, I don't feel like your fear-mongering regarding IP addresses is valid. Regardless if a user is on a residential or business connection no personal information can be obtained. Whois information on business IPs are the exact same as IPs aren't individually SWIPed with your personal address. All contact information would remain the ISPs regardless of personal or business connections. If you can link me evidence of such actions where an address was retrieved  that didn't involve social engineering I would love to take back this response.

I will be closing this issue soon.

**Comment by ghost** [comment]
From previous law cases, I know for a fact that businesses can have IP's linked to a business address. That business address may be a home address for some users/streamers. It was never fear-mongering and is a legitmate privacy issue. Even without a home address, a streamer/user is open to general location, DDoS and swatting. That information located in the token was never needed and it is obviously very frustrating that it was there in the first place when many extensions before had the code ready made to exclude it. I made the pull request to help you out, so thanks for calling me a fear-monger.

**Comment by captainswain** [comment]
> From previous law cases, I know for a fact that businesses can have IP's linked to a business address. That business address may be a home address for some users/streamers. It was never fear-mongering and is a legitmate privacy issue. Even without a home address, a streamer/user is open to general location, DDoS and swatting. That information located in the token was never needed and it is obviously very frustrating that it was there in the first place when many extensions before had the code ready made to exclude it. I made the pull request to help you out, so thanks for calling me a fear-monger.

Again link me to something that isn't circumstantial that supports your claims. I think you have a misconstrued understanding of how IPs work. People aren't going to get swatted purely off of the IP geolocation and generic ISP whois information. I don't disagree that it would be better from a privacy standpoint to remove the data (hence I merged your PR), but I personally felt that you brought the issue to attention in a very alarmist way.  Like I said in my previous comment, I really appreciate you putting in the work to create the PR. 

Sorry for offending you, I just kept seeing you spread misinformation on various forums about potential swatting, major breaches of privacy, and other nefarious activities that could be associated with the plugin.

**Comment by ghost** [comment]
> 
> 
> > From previous law cases, I know for a fact that businesses can have IP's linked to a business address. That business address may be a home address for some users/streamers. It was never fear-mongering and is a legitmate privacy issue. Even without a home address, a streamer/user is open to general location, DDoS and swatting. That information located in the token was never needed and it is obviously very frustrating that it was there in the first place when many extensions before had the code ready made to exclude it. I made the pull request to help you out, so thanks for calling me a fear-monger.
> 
> Again link me to something that isn't circumstantial that supports your claims. I think you have a misconstrued understanding of how IPs work. People aren't going to get swatted purely off of the IP geolocation and generic ISP whois information. I don't disagree that it would be better from a privacy standpoint to remove the data (hence I merged your PR), but I personally felt that you brought the issue to attention in a very alarmist way. Like I said in my previous comment, I really appreciate you putting in the work to create the PR.

You can literally look up any business with a dedicated line and you can find the information, it's not the same as domain whois. I'm not going to post a business on here to prove the point, as it's personal information, unlike domains. This part about swatting not being possible is simply not true, if you think the police can't get someones address from a personal or business IP, you are wrong. It's not alarmist to bring up a privacy issue, that can easily be fixed and that was ignored for weeks on end. You yourself agree that it is a privacy issue and are even implementing the fix. If you are so sure it's harmless, then don't fix it.

> 
> Sorry for offending you, I just kept seeing you spread misinformation on various forums about potential swatting, major breaches of privacy, and other nefarious activities that could be associated with the plugin.

I would love to know what various forums I have been spreading misinformation. I opened this issue on YOUR Github repo, to help you fix an issue that was reported by another user on Reddit. I even made the damn pull request to fix it. There is no misinformation. Linking a userID and an IP associated with it, is personal information, although you state you loose that information during the request before the proxy, it is still collected when the request is sent. I don't know if you have different laws, but linking the two and collecting it without storing it, is collecting personal information. Your privacy statements suggest you do not collect personal information. Any information I give as a response to someone has been truthful.

Instead of trying to make me out to be some kind of fear-monger or alarmist, or someone who doesn't know what they are talking about. Just fix the privacy issue, close the issue and be done with it. If you think it is not a privacy issue, don't fix it.

It's been 30 days since I opened this issue for you on Github, to bring your attention to it after seeing the post on Reddit, I even commented on that post to say we are trying to help you fix it, yet you make me out to be the bad guy here?

Like I said fix it, or don't. I won't be helping you anymore.

**Comment by pixeltris** [comment]
Just a little bump. You haven't actually pushed this to chrome / firefox stores yet. Any ETA on that?

**Comment by RockNRollGeek** [comment]
> Just a little bump. You haven't actually pushed this to chrome / firefox stores yet. Any ETA on that?

And now it's August 7th.... still no update in stores, yet someone else was able to create a firefox/chrome extension to block other twitch extensions that are snagging IP's, and get both approved and live almost instantly.

**Comment by nygellyndley** [comment]
Would I be right in saying this merged in May but was never actually pushed to the Firefox/Chrome store?

**Comment by MisterSheeple** [comment]
Bump on this. It's been over 1.5 years. I guarantee you this change is something that the extension's 24,920 Firefox users and 300,000+ Chrome users would want to take advantage of. Any chance an update could be pushed?
